File: AdventOfCode.Solutions.Rust/src/y2024/d01.rs
use crate::core::{Part, Solution};

pub fn solve(part: Part, input: String) -> Option<String> {
    match part {
        Part::P1 => Day01::solve_part_one(input),
        Part::P2 => Day01::solve_part_two(input),
    }
}

struct Day01;
impl Solution for Day01 {
    fn solve_part_one(input: String) -> Option<String> {
        let mut list1: Vec<i32> = vec![];
        let mut list2: Vec<i32> = vec![];

        input.lines().for_each(|l| {
            let mut parts = l.split_whitespace();
            if let (Some(a), Some(b)) = (parts.next(), parts.next()) {
                list1.push(a.parse::<i32>().unwrap());
                list2.push(b.parse::<i32>().unwrap());
            }
        });

        list1.sort();
        list2.sort();

        Some(
            list1
                .iter()
                .zip(list2.iter())
                .map(|(a, b)| (a - b).abs())
                .sum::<i32>()
                .to_string(),
        )
    }

    fn solve_part_two(input: String) -> Option<String> {
        let mut list1: Vec<i32> = vec![];
        let mut list2: Vec<i32> = vec![];

        input.lines().for_each(|l| {
            let mut parts = l.split_whitespace();
            if let (Some(a), Some(b)) = (parts.next(), parts.next()) {
                list1.push(a.parse::<i32>().unwrap());
                list2.push(b.parse::<i32>().unwrap());
            }
        });

        Some(
            list1
                .iter()
                .map(|a| a * list2.iter().filter(|&b| b == a).count() as i32)
                .sum::<i32>()
                .to_string(),
        )
    }
}

File: AdventOfCode.Solutions.Zig/src/y2024/d01.zig
const std = @import("std");

const core = @import("../core.zig");

pub fn solve(part: core.Part, input: []const u8) ?[*]u8 {
    return switch (part) {
        .Part1 => solvePartOne(input),
        .Part2 => solvePartTwo(input),
    };
}

pub fn solvePartOne(input: []const u8) ?[*]u8 {
    const allocator = std.heap.page_allocator;

    var list1: std.ArrayList(i32) = .empty;
    defer list1.deinit(allocator);

    var list2: std.ArrayList(i32) = .empty;
    defer list2.deinit(allocator);

    var iter = std.mem.tokenizeScalar(u8, input, '\n');
    while (iter.next()) |line| {
        var parts = std.mem.tokenizeScalar(u8, line, ' ');

        if (parts.next()) |a| {
            if (parts.next()) |b| {
                list1.append(allocator, try parseI32(a)) catch return null;
                list2.append(allocator, try parseI32(b)) catch return null;
            }
        }
    }

    std.mem.sort(i32, list1.items, {}, comptime std.sort.asc(i32));
    std.mem.sort(i32, list2.items, {}, comptime std.sort.asc(i32));

    var sum: u32 = 0;
    const count = list1.items.len;
    var i: usize = 0;
    while (i < count) : (i += 1) {
        const a = list1.items[i];
        const b = list2.items[i];
        sum += @intCast(@abs(a - b));
    }

    return core.toString(u32, &allocator, sum);
}

pub fn solvePartTwo(input: []const u8) ?[*]u8 {
    _ = input;
    return null;
}

fn parseI32(value: []const u8) !i32 {
    var result: i32 = 0;

    for (value) |ch| {
        const digit = ch - '0';
        result = result * 10 + digit;
    }

    return result;
}
